File: packages/encrypted-container/src/crypto.ts
import {
  argon2,
  createCipheriv,
  createDecipheriv,
  createHmac,
  hkdfSync,
  timingSafeEqual,
  webcrypto
} from "node:crypto";

import {
  BEC1_GCM_TAG_SIZE,
  BEC1_KEY_SIZE,
  BEC1_MAX_KDF_MEMORY_KIB,
  BEC1_MAX_KDF_PARALLELISM,
  BEC1_MAX_KDF_PASSES,
  BEC1_MIN_KDF_MEMORY_KIB,
  BEC1_MIN_KDF_PASSES
} from "./constants.js";
import { asAuthenticationError, Bec1Error } from "./errors.js";
import type { Bec1KdfParameters } from "./types.js";

const subtle = webcrypto.subtle;

export const validateKdfParameters = (kdf: Bec1KdfParameters): void => {
  if (
    kdf.algorithm !== "argon2id" ||
    kdf.version !== 0x13 ||
    kdf.tagLength !== BEC1_KEY_SIZE
  ) {
    throw new Bec1Error(
      "UNSUPPORTED_VERSION",
      "The container uses an unsupported password KDF."
    );
  }

  if (
    !Number.isInteger(kdf.memoryKiB) ||
    !Number.isInteger(kdf.passes) ||
    !Number.isInteger(kdf.parallelism) ||
    kdf.parallelism < 1 ||
    kdf.parallelism > BEC1_MAX_KDF_PARALLELISM ||
    kdf.passes < BEC1_MIN_KDF_PASSES ||
    kdf.passes > BEC1_MAX_KDF_PASSES ||
    kdf.memoryKiB < Math.max(8 * kdf.parallelism, BEC1_MIN_KDF_MEMORY_KIB) ||
    kdf.memoryKiB > BEC1_MAX_KDF_MEMORY_KIB
  ) {
    throw new Bec1Error(
      "UNSAFE_PARAMETERS",
      "The container declares unsafe Argon2id parameters."
    );
  }
};

export const deriveRecoveryKek = async (
  password: string,
  salt: Uint8Array,
  kdf: Bec1KdfParameters
): Promise<Buffer> => {
  validateKdfParameters(kdf);
  const normalizedPassword = password.normalize("NFC");
  for (let index = 0; index < normalizedPassword.length; index += 1) {
    const codeUnit = normalizedPassword.charCodeAt(index);
    if (codeUnit >= 0xd800 && codeUnit <= 0xdbff) {
      const next = normalizedPassword.charCodeAt(index + 1);
      if (next < 0xdc00 || next > 0xdfff) {
        throw new Bec1Error(
          "UNSAFE_PARAMETERS",
          "The password contains an unpaired Unicode surrogate."
        );
      }
      index += 1;
    } else if (codeUnit >= 0xdc00 && codeUnit <= 0xdfff) {
      throw new Bec1Error(
        "UNSAFE_PARAMETERS",
        "The password contains an unpaired Unicode surrogate."
      );
    }
  }
  const message = Buffer.from(normalizedPassword, "utf8");

  try {
    return await new Promise<Buffer>((resolve, reject) => {
      argon2(
        "argon2id",
        {
          message,
          nonce: salt,
          parallelism: kdf.parallelism,
          tagLength: kdf.tagLength,
          memory: kdf.memoryKiB,
          passes: kdf.passes
        },
        (error, derivedKey) => {
          message.fill(0);
          if (error) {
            reject(error);
            return;
          }
          resolve(Buffer.from(derivedKey));
        }
      );
    });
  } catch (error) {
    message.fill(0);
    throw error;
  }
};

export const deriveSubkey = (
  key: Uint8Array,
  salt: Uint8Array,
  info: Uint8Array
): Buffer => Buffer.from(hkdfSync("sha256", key, salt, info, BEC1_KEY_SIZE));

export const wrapKey = async (
  wrappingKey: Uint8Array,
  rawKey: Uint8Array
): Promise<Buffer> => {
  const kek = await subtle.importKey("raw", wrappingKey, "AES-KW", false, [
    "wrapKey"
  ]);
  const key = await subtle.importKey("raw", rawKey, "AES-GCM", true, [
    "encrypt",
    "decrypt"
  ]);
  return Buffer.from(await subtle.wrapKey("raw", key, kek, "AES-KW"));
};

export const unwrapKey = async (
  wrappingKey: Uint8Array,
  wrappedKey: Uint8Array
): Promise<Buffer> => {
  try {
    const kek = await subtle.importKey("raw", wrappingKey, "AES-KW", false, [
      "unwrapKey"
    ]);
    const key = await subtle.unwrapKey(
      "raw",
      wrappedKey,
      kek,
      "AES-KW",
      "AES-GCM",
      true,
      ["encrypt", "decrypt"]
    );
    return Buffer.from(await subtle.exportKey("raw", key));
  } catch (error) {
    throw asAuthenticationError(error);
  }
};

export const encryptAesGcm = (
  plaintext: Uint8Array,
  key: Uint8Array,
  nonce: Uint8Array,
  aad: Uint8Array
): Buffer => {
  const cipher = createCipheriv("aes-256-gcm", key, nonce, {
    authTagLength: BEC1_GCM_TAG_SIZE
  });
  cipher.setAAD(aad);
  const ciphertext = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  return Buffer.concat([ciphertext, cipher.getAuthTag()]);
};

export const decryptAesGcm = (
  ciphertextAndTag: Uint8Array,
  key: Uint8Array,
  nonce: Uint8Array,
  aad: Uint8Array
): Buffer => {
  if (ciphertextAndTag.byteLength < BEC1_GCM_TAG_SIZE) {
    throw asAuthenticationError(
      new Error("The authenticated ciphertext is truncated.")
    );
  }

  try {
    const ciphertext = Buffer.from(ciphertextAndTag).subarray(
      0,
      -BEC1_GCM_TAG_SIZE
    );
    const tag = Buffer.from(ciphertextAndTag).subarray(-BEC1_GCM_TAG_SIZE);
    const decipher = createDecipheriv("aes-256-gcm", key, nonce, {
      authTagLength: BEC1_GCM_TAG_SIZE
    });
    decipher.setAAD(aad);
    decipher.setAuthTag(tag);
    return Buffer.concat([decipher.update(ciphertext), decipher.final()]);
  } catch (error) {
    throw asAuthenticationError(error);
  }
};

export const descriptorMac = (
  descriptorBody: Uint8Array,
  key: Uint8Array
): Buffer => createHmac("sha256", key).update(descriptorBody).digest();

export const verifyDescriptorMac = (
  descriptorBody: Uint8Array,
  expected: Uint8Array,
  key: Uint8Array
): void => {
  const actual = descriptorMac(descriptorBody, key);
  if (
    actual.byteLength !== expected.byteLength ||
    !timingSafeEqual(actual, expected)
  ) {
    throw asAuthenticationError(
      new Error("The descriptor authentication code does not match.")
    );
  }
};

File: tools/bec1-interop-rust/src/crypto.rs
use aes_gcm::{
    aead::{Aead, KeyInit, Payload},
    aes::cipher::consts::U12,
    Aes256Gcm, Nonce,
};
use aes_kw::KekAes256;
use argon2::{Algorithm, Argon2, Params, Version};
use hkdf::Hkdf;
use hmac::{Hmac, KeyInit as HmacKeyInit, Mac};
use serde::Serialize;
use sha2::Sha256;
use unicode_normalization::UnicodeNormalization;
use zeroize::Zeroizing;

use crate::{
    constants::{
        KEY_SIZE, MAX_KDF_MEMORY_KIB, MAX_KDF_PARALLELISM, MAX_KDF_PASSES, MIN_KDF_MEMORY_KIB,
        MIN_KDF_PASSES,
    },
    error::{Bec1Error, Result},
};

#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct KdfParameters {
    pub algorithm: &'static str,
    pub version: u8,
    #[serde(rename = "memoryKiB")]
    pub memory_kib: u32,
    pub passes: u32,
    pub parallelism: u32,
    pub tag_length: u16,
}

impl KdfParameters {
    pub fn validate(&self) -> Result<()> {
        if self.version != 0x13 || self.tag_length != KEY_SIZE as u16 {
            return Err(Bec1Error::new(
                "UNSUPPORTED_VERSION",
                "The container uses an unsupported password KDF.",
            ));
        }
        if self.parallelism == 0
            || self.parallelism > MAX_KDF_PARALLELISM
            || self.passes < MIN_KDF_PASSES
            || self.passes > MAX_KDF_PASSES
            || self.memory_kib < MIN_KDF_MEMORY_KIB.max(8 * self.parallelism)
            || self.memory_kib > MAX_KDF_MEMORY_KIB
        {
            return Err(Bec1Error::unsafe_parameters(
                "The container declares unsafe Argon2id parameters.",
            ));
        }
        Ok(())
    }
}

pub fn normalize_password(password: &str) -> String {
    password.nfc().collect()
}

pub fn derive_recovery_kek(
    password: &str,
    salt: &[u8],
    kdf: &KdfParameters,
) -> Result<Zeroizing<[u8; KEY_SIZE]>> {
    kdf.validate()?;
    let normalized = Zeroizing::new(normalize_password(password));
    let parameters = Params::new(kdf.memory_kib, kdf.passes, kdf.parallelism, Some(KEY_SIZE))
        .map_err(|error| Bec1Error::unsafe_parameters(error.to_string()))?;
    let algorithm = Argon2::new(Algorithm::Argon2id, Version::V0x13, parameters);
    let mut output = Zeroizing::new([0_u8; KEY_SIZE]);
    algorithm
        .hash_password_into(normalized.as_bytes(), salt, output.as_mut())
        .map_err(|error| Bec1Error::unsafe_parameters(error.to_string()))?;
    Ok(output)
}

pub fn derive_subkey(key: &[u8], salt: &[u8], info: &[u8]) -> Result<Zeroizing<[u8; KEY_SIZE]>> {
    let hkdf = Hkdf::<Sha256>::new(Some(salt), key);
    let mut output = Zeroizing::new([0_u8; KEY_SIZE]);
    hkdf.expand(info, output.as_mut())
        .map_err(|_| Bec1Error::invalid("HKDF output length is invalid."))?;
    Ok(output)
}

pub fn wrap_key(wrapping_key: &[u8], raw_key: &[u8]) -> Result<Vec<u8>> {
    let kek = KekAes256::try_from(wrapping_key)
        .map_err(|_| Bec1Error::invalid("AES-KW key length is invalid."))?;
    kek.wrap_vec(raw_key)
        .map_err(|_| Bec1Error::authentication())
}

pub fn unwrap_key(wrapping_key: &[u8], wrapped_key: &[u8]) -> Result<Zeroizing<Vec<u8>>> {
    let kek = KekAes256::try_from(wrapping_key).map_err(|_| Bec1Error::authentication())?;
    let output = kek
        .unwrap_vec(wrapped_key)
        .map_err(|_| Bec1Error::authentication())?;
    if output.len() != KEY_SIZE {
        return Err(Bec1Error::authentication());
    }
    Ok(Zeroizing::new(output))
}

pub fn encrypt_aes_gcm(plaintext: &[u8], key: &[u8], nonce: &[u8], aad: &[u8]) -> Result<Vec<u8>> {
    if nonce.len() != 12 {
        return Err(Bec1Error::invalid("AES-GCM nonce length is invalid."));
    }
    let cipher =
        Aes256Gcm::new_from_slice(key).map_err(|_| Bec1Error::invalid("AES key is invalid."))?;
    let nonce = Nonce::<U12>::try_from(nonce)
        .map_err(|_| Bec1Error::invalid("AES-GCM nonce length is invalid."))?;
    cipher
        .encrypt(
            &nonce,
            Payload {
                msg: plaintext,
                aad,
            },
        )
        .map_err(|_| Bec1Error::authentication())
}

pub fn decrypt_aes_gcm(
    ciphertext_and_tag: &[u8],
    key: &[u8],
    nonce: &[u8],
    aad: &[u8],
) -> Result<Zeroizing<Vec<u8>>> {
    if nonce.len() != 12 || ciphertext_and_tag.len() < 16 {
        return Err(Bec1Error::authentication());
    }
    let cipher = Aes256Gcm::new_from_slice(key).map_err(|_| Bec1Error::authentication())?;
    let nonce = Nonce::<U12>::try_from(nonce).map_err(|_| Bec1Error::authentication())?;
    cipher
        .decrypt(
            &nonce,
            Payload {
                msg: ciphertext_and_tag,
                aad,
            },
        )
        .map(Zeroizing::new)
        .map_err(|_| Bec1Error::authentication())
}

pub fn descriptor_mac(body: &[u8], key: &[u8]) -> Result<[u8; 32]> {
    let mut mac = <Hmac<Sha256> as HmacKeyInit>::new_from_slice(key)
        .map_err(|_| Bec1Error::invalid("HMAC key is invalid."))?;
    mac.update(body);
    Ok(mac.finalize().into_bytes().into())
}

pub fn verify_descriptor_mac(body: &[u8], expected: &[u8], key: &[u8]) -> Result<()> {
    let mut mac = <Hmac<Sha256> as HmacKeyInit>::new_from_slice(key)
        .map_err(|_| Bec1Error::authentication())?;
    mac.update(body);
    mac.verify_slice(expected)
        .map_err(|_| Bec1Error::authentication())
}
